dashboard/app/models/user.rb

```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
     if !script_sections.empty?
       # if we have one or more sections matching this script id, we consider a stage hidden if all of those sections
       # hides the stage
-      script_sections.all? {|s| script_level.stage_hidden_for_section?(s.id)}
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extra a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  def get_hidden_ids(assign_id, hidden_stages = true)
+    # If we're a teacher, we want to go through each of our sections and return
+    # a mapping from section id to hidden stages in that section
+    if teacher?
+      hidden_by_section = {}
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
+      end
+      return hidden_by_section
```

Comment:
You've got two totally non-overlapping cases in teacher/student here.  Maybe break out private helpers `get_hidden_ids_for_teacher` `get_hidden_ids_for_student`?

Reply:
Agreed, this would be more readable.